packages/app/src/cli/models/app/app.ts

```diff
@@ -1,22 +1,102 @@
 import {AppErrors} from './loader.js'
 import {ExtensionInstance} from '../extensions/extension-instance.js'
+import {isType} from '../../utilities/types.js'
 import {zod} from '@shopify/cli-kit/node/schema'
 import {DotEnvFile} from '@shopify/cli-kit/node/dot-env'
 import {getDependencies, PackageManager, readAndParsePackageJson} from '@shopify/cli-kit/node/node-package-manager'
 import {fileRealPath, findPathUp} from '@shopify/cli-kit/node/fs'
 import {joinPath, dirname} from '@shopify/cli-kit/node/path'
+import {AbortError} from '@shopify/cli-kit/node/error'
 
-export const AppConfigurationSchema = zod.object({
+export const LegacyAppSchema = zod.object({
+  extension_directories: zod.array(zod.string()).optional(),
+  web_directories: zod.array(zod.string()).optional(),
+  scopes: zod.string().default(''),
+})
+
+export const AppSchema = zod.object({
   extension_directories: zod.array(zod.string()).optional(),
   web_directories: zod.array(zod.string()).optional(),
   name: zod.string().optional(),
   client_id: zod.string().optional(),
-  scopes: zod.string().default(''),
   application_url: zod.string().optional(),
   redirect_url_allowlist: zod.array(zod.string()).optional(),
   requested_access_scopes: zod.array(zod.string()).optional(),
+  scopes: zod.string().optional(),
 })
 
+export const AppConfigurationSchema = zod.union([LegacyAppSchema, AppSchema])
+
+export interface SchemaValidationOptions {
+  strict?: boolean
+}
+
+/**
+ * Check whether a shopify.app.toml schema is valid against the legacy schema definition.
+ * @param item - the item to validate
+ * @param strict - whether to allow keys not defined in the schema
+ */
+export function isLegacyAppSchema(
+  item: unknown,
+  options?: SchemaValidationOptions,
+): item is zod.infer<typeof LegacyAppSchema> {
+  if (options?.strict) return isType(LegacyAppSchema.strict(), item)
+  return isType(LegacyAppSchema, item)
+}
+
+/**
+ * Check whether a shopify.app.toml schema is valid against the current schema definition.
+ * @param item - the item to validate
+ * @param strict - whether to allow keys not defined in the schema
+ */
+export function isCurrentAppSchema(
+  item: unknown,
+  options?: SchemaValidationOptions,
+): item is zod.infer<typeof AppSchema> {
+  if (options?.strict) return isType(AppSchema.strict(), item)
+  return isType(AppSchema, item)
+}
+
+/**
+ * Check whether a shopify.app.toml schema is valid.
+ * @param item - the item to validate
+ * @param strict - whether to allow keys not defined in the schema
+ */
+export function isValidAppSchema(
+  item: unknown,
+  options?: SchemaValidationOptions,
+): item is zod.infer<typeof AppConfigurationSchema> {
+  return isLegacyAppSchema(item, {strict: options?.strict}) || isCurrentAppSchema(item, {strict: options?.strict})
+}
+
+/**
+ * Get scopes from a given app.toml config file.
+ * @param config - a configuration file
+ */
+export function getAppScopes(config: unknown) {
```

Comment:
I think we'll ant to talk about future formatting of scopes since we have 2 options for types now, but for now we are still treating implementation as expecting a string